core/src/main/java/org/apache/calcite/sql/SqlSelect.java

```diff
@@ -288,4 +297,133 @@ public boolean hasWhere() {
+        if (operands.size() == 2 && (operands.get(0) instanceof SqlIdentifier
+            || operands.get(0) instanceof SqlLiteral)) {
+          // No need to replace if the alias is a simple column
+          // rename or a simple name for a literal.
+          continue;
+        }
+        possibleAliases.add(aliasIdentifier.names);
+      }
+    }
+
+    // If we don't have any aliases we won't need this rewrite
+    if (possibleAliases.isEmpty()) {
+      return this;
+    }
+    // Check if the Where clause uses this alias anywhere. If so
+    // we need to replace.
+    ArrayDeque<SqlNode> nodeQueue = new ArrayDeque<>();
+    nodeQueue.add(requireNonNull(where, "where"));
+    boolean needsReplacement = false;
+    while (!nodeQueue.isEmpty()) {
+      SqlNode node = nodeQueue.pop();
+      if (node instanceof SqlCall) {
+        SqlCall callNode = (SqlCall) node;
+        List<SqlNode> operands = callNode.getOperandList();
+        for (SqlNode operand: operands) {
+          if (operand != null) {
+            nodeQueue.add(operand);
+          }
+        }
+      } else if (node instanceof SqlIdentifier) {
+        SqlIdentifier identifier = (SqlIdentifier) node;
+        if (possibleAliases.contains(identifier.names)) {
+          needsReplacement = true;
+          break;
+        }
+      }
+    }
+    // We didn't find any aliases that need replacement
+    // we don't need to rewrite this Node.
+    if (!needsReplacement) {
+      return this;
+    }
+
+    // The rewrite is required, so we need to generate new Select statements from the
+    // new node lists. Generate an inner select without the where we need to update
+    // the inner select to include any column in the FROM.
+    List<@Nullable SqlNode> innerNodes = new ArrayList<>();
+    List<@Nullable SqlNode> outerNodes = new ArrayList<>();
+
+    // Also generate what the new select statements
+    // would be at this time.
+    for (int i = 0; i < selectList.size(); i++) {
+      SqlNode selectVal = selectList.get(i);
+      SqlIdentifier aliasIdentifier;
+      SqlNode innerVal;
+      if (selectVal instanceof SqlBasicCall && selectVal.getKind() == SqlKind.AS) {
+        // We have found an alias
+        innerVal = selectVal;
+        SqlBasicCall innerAlias = (SqlBasicCall) selectVal;
+        aliasIdentifier = (SqlIdentifier) innerAlias.getOperandList()
+            .get(innerAlias.operandCount() - 1);
+      } else if (selectVal instanceof SqlIdentifier) {
+        // If we have just found an identifier we don't want to generate an alias.
+        innerVal = selectVal;
+        aliasIdentifier = (SqlIdentifier) selectVal;
+      } else {
+        // We need to generate a new alias with an internal name.
+        // TODO: Check this for uniqueness???
+        aliasIdentifier = new SqlIdentifier(
+            String.format(Locale.ROOT,
+            "$TEMP_COLUMN%d", i), selectVal.getParserPosition());
```

Comment:
Isn't this fine? I thought the reason that calcite used $colname was that it was impossible for a column identifier to parse with that syntax.

Reply:
Never mind, 'select a as $test' works fine for me. Can we maybe add some prefix like CALICITE_INTERNAL_TEMP_COLUMN? TEMP_COLUMN is a generic enough name that I could potentially see some sort of conflict.

Reply:
I like starting with `$` since calcite internal name use `$`. I the idea of starting with `$BODO_INTERNAL_TEMP_COLUMN`.